rtrlib/rtr_mgr.c

```diff
@@ -431,17 +525,155 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
 			    struct spki_record **result,
 			    unsigned int *result_count)
 {
-	return spki_table_get_all(config->groups[0].sockets[0]->spki_table,
+	return spki_table_get_all(config->spki_table,
 				  asn, ski, result, result_count);
 }
 
 void rtr_mgr_stop(struct rtr_mgr_config *config)
 {
+	struct rtr_mgr_group_node *group_node;
+
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
 	MGR_DBG1("rtr_mgr_stop()");
-	for (unsigned int i = 0; i < config->len; i++) {
-		for (unsigned int j = 0; j < config->groups[i].sockets_len; j++)
-			rtr_stop(config->groups[i].sockets[j]);
+	while (node) {
+		group_node = node->data;
+		for (unsigned int j = 0; j < group_node->group->sockets_len;
+		     j++) {
+			rtr_stop(group_node->group->sockets[j]);
+		}
+	node = node->next;
+	}
+	pthread_mutex_unlock(&config->mutex);
+}
+
+int rtr_mgr_add_group(struct rtr_mgr_config *config,
+		      const struct rtr_mgr_group *group)
+{
+	unsigned int refresh_iv = 3600;
+	unsigned int retry_iv = 600;
+	unsigned int expire_iv = 7200;
+
+	pthread_mutex_lock(&config->mutex);
+
+	struct rtr_mgr_group_node *gnode;
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+		gnode = node->data;
+		if (gnode->group->preference == group->preference) {
+			MGR_DBG1("Group with preference value already exists!");
+			pthread_mutex_unlock(&config->mutex);
+			return RTR_INVALID_PARAM;
```

Comment:
instead of having to enable debugging to find out the  cause it would be great to return an error code that describes the reason more specific